### PR TITLE
Add controller typespecs and make function definitions more strict

### PR DESCRIPTION
**Pull request by gradenb**
Please let me know if any of these look incorrect or if there is a better way of writing any of them.

### CODE REVIEW

**Comment by gradenb** [comment]
I also added additional pattern matching to public function definitions in Controller.ex.
The idea with that is to prevent odd behavior similar to that seen here:
https://github.com/phoenixframework/phoenix/pull/2210

**Comment by gradenb** [comment]
Maybe it's worth it to alias Plug.Conn with this many uses...?

**Comment by gradenb** [comment]
Are changes like this welcome? I imagine that the type specs are wanted on publicly documented API. I can drop the latter commit if we are worried that it will cause unintended behavior. I'm interested in getting involved in the elixir community, but realize I'm a newcomer. I'm certainly open to constructive criticism if these changes aren't perfect.

File: lib/phoenix/controller.ex [hunk 1]
```diff
@@ -224,6 +224,7 @@ defmodule Phoenix.Controller do
         end
       end
   """
+  @spec action_fallback(module) :: atom
```
Comment (ericmj) [inline lib/phoenix/controller.ex]: Macros shouldn't be specced.

**Comment by chrismccord** [comment]
> Are changes like this welcome?

Thanks! Our approach is to spec public apis as we go. We avoid these kinds of contributions because it can introduce churn and conflicts b/w releases, but we much appreciate your contributions and the ones we've merged previously! :)

**Pull request closed without merging**